### ISSUE TITLE
swc code generation is slower than other implementations

### ISSUE THREAD

**Issue by connor4312** [swc-project/swc#695]
**Context**

I am looking for ways to transform/beautify large minified scripts for some new functionality in the VS Code JavaScript debugger. Right now we use js-beautify, but it doesn't generate sourcemaps and we have a rather naive, somewhat buggy way to create them after the fact.

I wanted to look into swc as an alternative, ideally writing a super fast sourcemap-aware beautifier I could throw into .wasm. However I was surprised the codegen performance, especially with sourcemaps enabled, wasn't so great.

**Describe the bug**

Measuring beautification on [this large JavaScript bundle](https://mixer.com/_latest/bundles/main.en-US-a8c0cdbf4d5dc8679138.bundle.js?71316da4d40c63c9fabc), these are the timings I get:

```
                  | w/ sourcemaps | only generation | 
      js-beautify |               |       1967      |
              swc |     25208     |       3509      |
            babel |      5128     |       5470      |
esprima+escodegen |     10256     |       1632      |
```

 - js-beautify is passed through its included CLI
 - swc is parsing and then emitting source into a vec, built with cargo build --release
 - esprima+escodegen is similarly parsing the source with esprima and then writing it to a string with escodegen (source in gist)
 - babel is the same thing with @babel/parser and @babel/generator

**Input code**

https://gist.github.com/connor4312/5860dbc0970ad44ba30d91b5aa2812f0

**Expected behavior**

I was hoping swc would be much faster than the JavaScript implementations!

**Version**

```
[dependencies]
sourcemap = "4.1"
swc_common = "0.5"
swc_ecma_ast = "0.17"
swc_ecma_parser = "0.19"
swc_ecma_codegen = "0.16"
```

**Comment by kdy1** [comment]
# Previous

```
emit_colors ... bench:      23,601 ns/iter (+/- 15,849) = 49 MB/s
```


# After https://github.com/kdy1/swc/commit/0e30f3b499b87c147319ae870b148e72767d25c8

```
test emit_colors ... bench:      13,659 ns/iter (+/- 450) = 86 MB/s
test emit_large  ... bench:     112,836 ns/iter (+/- 16,139) = 45 MB/s
```


# After https://github.com/kdy1/swc/commit/6c5cc5da4709d6f8d6aa2e8a2065a783a8340ad5

```
test emit_colors ... bench:      13,501 ns/iter (+/- 10,828) = 87 MB/s
test emit_large  ... bench:     109,805 ns/iter (+/- 2,409) = 46 MB/s
```

As swc's code generator is designed to 

 - support rich error reporting
 - allow emitting merged modules (for the bundler I'll write)

it has some performance drop. However, I think I can make it faster than other js implementations.
 
I'll investigate further.

**Comment by connor4312** [comment]
I was running the binary here normally--not in wasm.

**Comment by kdy1** [comment]
Oh.. I got it.

Possible candidates for bottleneck would be

 - source code analysis for multibyte char and line information (`cm.new_source_file`)
 - parser (`parser.parse_module()`)
 - codegen

My guess is the parser. There are some low-hanging fruits.

**Comment by connor4312** [comment]
For my case I'm not sure the parser was the issue--passing `Some(&mut src_map_builder)` versus `None` to the `JsWriter` made the difference between a 3.5 second runtime and 25 second runtime.

**Comment by kdy1** [comment]
@connor4312 This is fixed in `swc_ecma_parser@0.20.0` and `swc_ecma_codegen@0.17.0`.  (and I verified on my linux device)

**Comment by kdy1** [comment]
I've published `swc_ecma_codegen@v0.18.0`, which fixes this issue.

**Comment by connor4312** [comment]
Wow, that made a huge difference! 25 seconds down to 2100ms on my machine. Fantastic work, responsiveness 👏
